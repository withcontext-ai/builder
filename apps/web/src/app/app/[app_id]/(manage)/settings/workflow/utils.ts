import { safeParse } from '@/lib/utils'
import { TreeItem } from '@/components/dnd/types'

import { WorkflowItem } from './type'

export function taskToApiFormatter(task: WorkflowItem) {
  const { key, type, subType } = task
  const chain = safeParse(task.formValueStr, {})
  if (chain.data) {
    chain.datasets = chain.data.datasets || []
    delete chain.data
  }
  return {
    key: `${type}-${key}`,
    chain_type: subType,
    ...chain,
  }
}

export function formatTreeWithData(
  tree: TreeItem[],
  data: WorkflowItem[],
  result: WorkflowItem[] = []
): WorkflowItem[] {
  for (const t of tree) {
    const task = data.find((d) => d.id === t.id)
    if (task) result.push(task)

    if (t.children && t.children.length > 0) {
      formatTreeWithData(t.children, data, result)
    }
  }

  return result
}
