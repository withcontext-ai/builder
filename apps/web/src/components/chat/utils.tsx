import { type Message as RawMessage } from 'ai'

import { Message as MessageSchema } from '@/db/messages/schema'

import { ChatMessage, EventMessage } from './types'

interface Message extends MessageSchema {
  role: RawMessage['role']
}

export const chatMessagesFilter = (m: Message) => m.type === 'chat'

export const eventMessagesFilter = (m: Message) => m.type === 'event'

export const chatMessagesFormatter = (m: Message) =>
  ({
    id: m.short_id,
    createdAt: m.created_at,
    role: m.role,
    content: m.content,
    type: 'chat',
    feedback: m.feedback,
    feedback_content: m.feedback_content,
    meta: {
      latency: m.latency,
      token: {
        total_tokens: m.total_tokens,
      },
      raw: m.raw,
    },
  }) as ChatMessage

export const eventMessagesFormatter = (m: Message) =>
  ({
    id: m.short_id,
    createdAt: m.created_at,
    role: m.role,
    content: m.content,
    type: 'event',
    eventType: m.event_type,
    call_duration: m.call_duration,
  }) as EventMessage

export const messageFormatter = (m: Message) => {
  if (m.type === 'chat') return chatMessagesFormatter(m)
  if (m.type === 'event') return eventMessagesFormatter(m)
  return null
}

export const messagesBuilder = (messages: Partial<MessageSchema>[]) => {
  const result = []
  for (const m of messages) {
    if (m.type === 'chat') {
      result.push({ ...m, role: 'user', content: m.query })
      if (m.answer) {
        result.push({ ...m, role: 'assistant', content: m.answer })
      }
    } else {
      result.push(m)
    }
  }
  return result as Message[]
}

export const keyBuilder = (m: RawMessage) => {
  if (m.role) return `${m.id}-${m.role}`
  return `${m.id}`
}
