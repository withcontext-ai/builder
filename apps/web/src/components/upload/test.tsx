'use client'

import { useState } from 'react'

import { UploadFile, UploadProps } from './type'
import Upload from './upload'

const TestUpload = (props: UploadProps) => {
  const [files, setFiles] = useState<UploadFile[]>([])

  const handleRemove = (file: UploadFile) => {
    const data = files?.filter((item) => item?.uid !== file?.uid)
    setFiles(data)
  }
  return (
    <Upload
      fileList={files}
      handleFiles={(files) => setFiles([...files])}
      onRemove={handleRemove}
      // 自定义请求方式，这里在onChange阶段调用的Google cloud，不处理会使用rc-upload的请求方式多发送一次请求
      customRequest={() => {}}
      {...props}
    />
  )
}

export default TestUpload
