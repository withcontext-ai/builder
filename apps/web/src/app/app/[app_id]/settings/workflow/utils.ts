import { safeParse } from '@/lib/utils'

import { WorkflowItem } from './type'

export function taskToApiFormatter(task: WorkflowItem) {
  const { key, type, subType } = task
  const chain = safeParse(task.formValueStr, {})
  return {
    key: `${type}-${key}`,
    chain_type: subType,
    ...chain,
  }
}
