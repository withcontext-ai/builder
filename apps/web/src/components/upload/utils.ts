import axios, { CancelTokenSource } from 'axios'

import { nanoid } from '@/lib/utils'

import type {
  InternalUploadFile,
  RcFile,
  UploadFile,
  UploadFileStatus,
} from './type'
import { listPropsInterface } from './type'

export function file2Obj(file: RcFile): InternalUploadFile {
  return {
    ...file,
    lastModified: file.lastModified,
    name: file.name,
    size: file.size,
    type: file.type,
    uid: file.uid,
    percent: 0,
    originFileObj: file,
  }
}

/** Upload fileList. Replace file if exist or just push into it. */
export function updateFileList(
  file: UploadFile,
  fileList: (UploadFile | Readonly<UploadFile>)[]
) {
  const nextFileList = [...fileList]
  const fileIndex = nextFileList.findIndex(({ uid }) => uid === file.uid)
  if (fileIndex === -1) {
    nextFileList.push(file)
  } else {
    nextFileList[fileIndex] = file
  }
  return nextFileList
}

export function getFileItem(
  file: RcFile,
  fileList: (UploadFile | Readonly<UploadFile>)[]
) {
  const matchKey = file.uid !== undefined ? 'uid' : 'name'
  return fileList.filter((item) => item[matchKey] === file[matchKey])[0]
}

export function removeFileItem(
  file: UploadFile,
  fileList: (UploadFile | Readonly<UploadFile>)[]
) {
  const matchKey = file.uid !== undefined ? 'uid' : 'name'
  const removed = fileList.filter((item) => item[matchKey] !== file[matchKey])
  if (removed.length === fileList.length) {
    return null
  }
  return removed
}

export const checkShowIcon = (listProps: boolean | listPropsInterface) => {
  if (typeof listProps === 'object') {
    return { show: true, ...listProps }
  } else {
    return { show: listProps }
  }
}

export const checkType = (file: UploadFile) => {
  if (file?.type !== 'application/pdf' || file?.url?.includes('.pdf')) {
    return 'isPDF'
  }
  return 'isImage'
}

export const getBase64 = (img: RcFile, callback: (url: string) => void) => {
  const reader = new FileReader()
  reader.addEventListener('load', () => callback(reader.result as string))
  reader.readAsDataURL(img)
}

const changeCurrentFile = async (
  file: UploadFile,
  mergedFileList: UploadFile[],
  setMergedFileList?: (files: UploadFile<any>[]) => void
) => {
  const index = mergedFileList?.indexOf(
    // @ts-ignore
    (item: { uid: any }) => item?.uid === file?.uid
  )

  if (index !== -1) {
    mergedFileList[index] = file
  }
  setMergedFileList?.(mergedFileList)
}

const handleSuccess = ({
  mergedFileList,
  fileType,
  onChangeFileList,
}: {
  mergedFileList: UploadFile[]
  fileType?: string
  onChangeFileList?: (files: FileProps[]) => void
}) => {
  const success = mergedFileList?.filter(
    (item) => item?.status === 'success' && item?.url
  )
  const data = success?.reduce((m: FileProps[], item: UploadFile) => {
    m.push({
      url: item?.url || '',
      uid: nanoid(),
      type: fileType || item?.type,
      name: item?.name,
    })
    return m
  }, [])
  onChangeFileList?.(data)
}

export const uploadFile = async ({
  file,
  mergedFileList,
  controller,
  source,
  onChangeFileList,
  setMergedFileList,
  setIsUploading,
  fileType,
}: {
  file: UploadFile
  mergedFileList: UploadFile<any>[]
  fileType?: string
  controller?: AbortController
  source?: CancelTokenSource
  onChangeFileList?: (files: FileProps[]) => void
  setMergedFileList?: (files: UploadFile<any>[]) => void
  setIsUploading: (s: boolean) => void
}) => {
  setIsUploading(true)
  if (!file) return
  file.status = 'uploading'
  file.percent = 10
  await changeCurrentFile(file, mergedFileList, setMergedFileList)
  const filename = encodeURIComponent(file?.name || '')
  const res = await fetch(`/api/upload-url/gcp?filename=${filename}`)
  const { success, data } = await res.json()
  if (!success) {
    file.status = 'error'
    setIsUploading(false)
    await changeCurrentFile(file, mergedFileList, setMergedFileList)
  }

  const { upload_url, upload_fields, file_url } = data as {
    provider: string
    upload_url: string
    upload_fields: { [key: string]: string }
    file_url: string
    file_path: string
  }
  const formData = new FormData()
  Object.entries({ ...upload_fields, file: file?.originFileObj }).forEach(
    ([key, value]) => {
      // @ts-ignore
      formData.append(key, value)
    }
  )

  axios
    .post(upload_url, formData, {
      signal: controller?.signal,
      cancelToken: source?.token,
      onUploadProgress: async (progressEvent) => {
        file.status = 'uploading'
        const { progress = 0.1 } = progressEvent
        file.percent = progress * 100
        await changeCurrentFile(file, mergedFileList, setMergedFileList)
      },
    })
    .then(async () => {
      file.status = 'success'
      file.url = file_url
      setIsUploading(false)
      handleSuccess({ mergedFileList, onChangeFileList, fileType })
      await changeCurrentFile(file, mergedFileList, setMergedFileList)
    })
    .catch((error) => {
      if (axios.isCancel(error)) {
        console.log('Request canceled', error.message)
      }
      setIsUploading(false)
      file.status = 'error'
      console.error(error)
    })
}

export const stringUrlToFile = (url: string) => {
  const status: UploadFileStatus = 'success'
  return url
    ? [
        {
          url,
          name: '',
          uid: nanoid(),
          status,
        },
      ]
    : []
}

export type FileProps = {
  name: string
  url: string
  uid?: string
  type?: string
}
export const changeToUploadFile = (
  value: Array<FileProps | string> | string
) => {
  const isFiles = Array.isArray(value)
  if (isFiles) {
    const data = value?.reduce((m: UploadFile[], item: FileProps | string) => {
      const isImage = typeof item === 'string'
      const status: UploadFileStatus = 'success'
      const cur = isImage
        ? stringUrlToFile(item)[0]
        : { ...item, status, uid: nanoid() }
      m.push(cur)
      return m
    }, [])
    return data
  } else {
    return stringUrlToFile(value)
  }
}
