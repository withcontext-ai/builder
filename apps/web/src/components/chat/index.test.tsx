'use client'

import { useEffect } from 'react'
import { cleanup, render, waitFor } from '@testing-library/react'
import user from '@testing-library/user-event'
import { nanoid } from 'nanoid'
import { afterEach, beforeAll, describe, expect, test, vi } from 'vitest'

import Chat from '@/components/chat/page'

import { commonTestWrapper } from '../../../test/utils'

export const testVideo = (video: boolean) => {
  return JSON.stringify([
    { formValueStr: JSON.stringify({ enable_video_interaction: video }) },
  ])
}

describe('Chat', () => {
  global.fetch = mockChat()
  function mockChat() {
    const mockMessages = [
      ['hello', 'world', 'thisis', 'atest'],
      ['hello', 'world', 'message2'],
    ]
    let i = 0
    return vi.fn().mockImplementation(
      (path: string) =>
        new Promise((res) => {
          if (path === '/api/chat') {
            const stream = new ReadableStream({
              start: async (controller) => {
                const encoder = new TextEncoder()
                if (i >= mockMessages.length) {
                  i = 0
                }
                const mockMessage = mockMessages[i++]
                await Promise.all(
                  mockMessage.map(
                    (m, i) =>
                      new Promise<void>((res) => {
                        setTimeout(() => {
                          try {
                            controller.enqueue(encoder.encode(m))
                          } catch {}
                          res()
                        }, i * 500)
                      })
                  )
                )
                controller.close()
              },
            })
            res(new Response(stream))
          }
        })
    )
  }

  beforeAll(() => {
    vi.mock('@clerk/nextjs', () => ({
      useUser: () => ({
        firstName: 'test',
        lastName: 'user',
        fullName: 'test user',
      }),
    }))
    vi.mock('@/hooks/use-subscribe', () => ({
      __esModule: true,
      default: ({ onAdd, enabled }: any) => {
        // eslint-disable-next-line react-hooks/rules-of-hooks
        useEffect(() => {
          if (enabled) {
            setTimeout(() => {
              onAdd({
                eventType: 'call.created',
                id: nanoid(),
              })
            }, 500)
          }
          // eslint-disable-next-line react-hooks/exhaustive-deps
        }, [])
      },
    }))
  })

  afterEach(() => {
    cleanup()
  })

  test('send message and regenerate', async () => {
    const component = render(
      commonTestWrapper(
        <Chat
          workflow={[]}
          app={{
            name: 'chat-app',
            short_id: 'YhTq4Xx29aDZ',
            icon: '',
            enable_video_interaction: false,
            opening_remarks: 'yo',
            published_workflow_data_str: testVideo(false),
          }}
          mode="live"
          session={{
            api_session_id: '21486acbbd393f8a6131a9009d6aae4d',
            short_id: 'srQuAKvgZR7W',
            name: 'test-chat-session',
          }}
          initialMessages={[
            {
              id: '1',
              content: 'hello1',
              role: 'user',
              type: 'chat',
              createdAt: new Date('2023-01-01 00:00:00'),
            },
            {
              id: '2',
              content: 'hello2',
              role: 'assistant',
              type: 'chat',
              createdAt: new Date('2023-01-01 00:00:01'),
            },
            {
              id: '3',
              content: 'hello3',
              role: 'user',
              type: 'chat',
              createdAt: new Date('2023-01-01 00:00:02'),
            },
            {
              id: '4',
              content: 'hello4',
              role: 'assistant',
              type: 'chat',
              createdAt: new Date('2023-01-01 00:00:03'),
            },
          ]}
        />
      )
    )
    const { asFragment, queryByTestId } = component
    expect(asFragment()).toMatchSnapshot()
    const sendButton = queryByTestId('send')!
    const input = queryByTestId('input')!

    expect(queryByTestId('send')).toBeInTheDocument()

    expect(sendButton).toBeDisabled()

    await user.click(input)

    await user.keyboard('hello, to test the textarea keypress')

    expect(sendButton).not.toBeDisabled()

    await user.click(sendButton)

    expect(input.textContent).toBe('')

    expect(sendButton).toBeDisabled()

    await waitFor(() => {
      expect(queryByTestId('stop')!).toBeInTheDocument()
    })
    await waitFor(
      () => {
        expect(
          component.queryByText('helloworldthisisatest')
        ).toBeInTheDocument()
      },
      {
        timeout: 10000,
      }
    )

    await waitFor(() => {
      const regen = component.queryByText('Regenerate response')!
      expect(regen).toBeInTheDocument()
      user.click(regen)
    })

    await waitFor(() => {
      const stop = queryByTestId('stop')!
      expect(stop).toBeInTheDocument()
      user.click(stop)
    })

    await waitFor(() => {
      const regen = component.queryByText('Regenerate response')!
      expect(regen).toBeInTheDocument()
    })
  })

  test('accept event', async () => {
    const open = vi.spyOn(window, 'open').mockImplementation(() => null)

    const component = render(
      commonTestWrapper(
        <Chat
          workflow={[]}
          app={{
            name: 'chat-app',
            short_id: 'YhTq4Xx29aDZ',
            icon: '',
            enable_video_interaction: true,
            opening_remarks: 'yo',
            published_workflow_data_str: testVideo(true),
          }}
          mode="live"
          session={{
            api_session_id: '21486acbbd393f8a6131a9009d6aae4d',
            short_id: 'srQuAKvgZR7W',
            name: 'test-chat-session',
          }}
        />
      )
    )
    await waitFor(async () => {
      const dialog = component.getByTestId('video-call-confirm-dialog')
      expect(dialog).toBeInTheDocument()
      await user.click(component.getByTestId('accept-video-call'))
      expect(open).toBeCalled()
      expect(dialog).not.toBeInTheDocument()
    })
  })

  test('decline event', async () => {
    const component = render(
      commonTestWrapper(
        <Chat
          workflow={[]}
          app={{
            name: 'chat-app',
            short_id: 'YhTq4Xx29aDZ',
            icon: '',
            enable_video_interaction: true,
            opening_remarks: 'yo',
            published_workflow_data_str: testVideo(true),
          }}
          mode="live"
          session={{
            api_session_id: '21486acbbd393f8a6131a9009d6aae4d',
            short_id: 'srQuAKvgZR7W',
            name: 'test-chat-session',
          }}
        />
      )
    )
    await waitFor(async () => {
      const dialog = component.getByTestId('video-call-confirm-dialog')
      expect(dialog).toBeInTheDocument()
      await user.click(component.getByTestId('decline-video-call'))
      expect(dialog).not.toBeInTheDocument()
      expect(component.getByText('Call Declined')).toBeInTheDocument()
    })
  })

  test('cancel event', async () => {
    vi.mock('usehooks-ts', () => ({
      ...vi.importActual('usehooks-ts'),
      useCountdown: () => [
        0,
        {
          startCountdown: () => {},
        },
      ],
    }))
    const component = render(
      commonTestWrapper(
        <Chat
          workflow={[]}
          app={{
            name: 'chat-app',
            short_id: 'YhTq4Xx29aDZ',
            icon: '',
            enable_video_interaction: true,
            opening_remarks: 'yo',
            published_workflow_data_str: testVideo(true),
          }}
          mode="live"
          session={{
            api_session_id: '21486acbbd393f8a6131a9009d6aae4d',
            short_id: 'srQuAKvgZR7W',
            name: 'test-chat-session',
          }}
        />
      )
    )
    await waitFor(async () => {
      const dialog = component.getByTestId('video-call-confirm-dialog')
      expect(dialog).toBeInTheDocument()
    })
    await waitFor(async () => {
      const dialog = component.queryByTestId('video-call-confirm-dialog')
      expect(dialog).not.toBeInTheDocument()
      expect(component.getAllByText('Call Canceled')).toBeTruthy()
    })
  })

  test('stop when generating', async () => {
    const component = render(
      commonTestWrapper(
        <Chat
          workflow={[]}
          app={{
            name: 'chat-app',
            short_id: 'YhTq4Xx29aDZ',
            icon: '',
            enable_video_interaction: false,
            opening_remarks: 'yo',
            published_workflow_data_str: testVideo(false),
          }}
          mode="live"
          session={{
            api_session_id: '21486acbbd393f8a6131a9009d6aae4d',
            short_id: 'srQuAKvgZR7W',
            name: 'test-chat-session',
          }}
          initialMessages={[
            {
              id: '1',
              content: 'hello1',
              role: 'user',
              type: 'chat',
              createdAt: new Date('2023-01-01 00:00:00'),
            },
            {
              id: '2',
              content: 'hello2',
              role: 'assistant',
              type: 'chat',
              createdAt: new Date('2023-01-01 00:00:01'),
            },
            {
              id: '3',
              content: 'hello3',
              role: 'user',
              type: 'chat',
              createdAt: new Date('2023-01-01 00:00:02'),
            },
            {
              id: '4',
              content: 'hello4',
              role: 'assistant',
              type: 'chat',
              createdAt: new Date('2023-01-01 00:00:03'),
            },
          ]}
        />
      )
    )

    const sendButton = component.queryByTestId('send')!
    const input = component.queryByTestId('input')!

    await user.click(input)

    await user.keyboard('hello, to test the textarea keypress2')

    await user.click(sendButton)

    await waitFor(async () => {
      expect(component.queryByText('hello')).toBeInTheDocument()
      await user.click(component.queryByTestId('stop')!)
      expect(component.queryByText('hello')).toBeInTheDocument()
      expect(
        component.queryByText('hello, to test the textarea keypress2')
      ).toBeInTheDocument()
    })
  })

  test('stop before generating', async () => {
    const component = render(
      commonTestWrapper(
        <Chat
          workflow={[]}
          app={{
            name: 'chat-app',
            short_id: 'YhTq4Xx29aDZ',
            icon: '',
            enable_video_interaction: false,
            opening_remarks: 'yo',
            published_workflow_data_str: testVideo(false),
          }}
          mode="live"
          session={{
            api_session_id: '21486acbbd393f8a6131a9009d6aae4d',
            short_id: 'srQuAKvgZR7W',
            name: 'test-chat-session',
          }}
          initialMessages={[
            {
              id: '1',
              content: 'hello1',
              role: 'user',
              type: 'chat',
              createdAt: new Date('2023-01-01 00:00:00'),
            },
            {
              id: '2',
              content: 'hello2',
              role: 'assistant',
              type: 'chat',
              createdAt: new Date('2023-01-01 00:00:01'),
            },
            {
              id: '3',
              content: 'hello3',
              role: 'user',
              type: 'chat',
              createdAt: new Date('2023-01-01 00:00:02'),
            },
            {
              id: '4',
              content: 'hello4',
              role: 'assistant',
              type: 'chat',
              createdAt: new Date('2023-01-01 00:00:03'),
            },
          ]}
        />
      )
    )

    const sendButton = component.queryByTestId('send')!
    const input = component.queryByTestId('input')!

    await user.click(input)

    await user.keyboard('testing stop before generating')

    await user.click(sendButton)

    await user.click(component.queryByTestId('stop')!)
    expect(
      component.queryByText('testing stop before generating')
    ).toBeInTheDocument()
  })
})
