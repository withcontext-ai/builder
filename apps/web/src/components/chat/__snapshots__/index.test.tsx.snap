// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chat > send message and regenerate 1`] = `
<DocumentFragment>
  <div
    class="relative h-full w-full"
  >
    <div
      class="flex h-full w-full flex-col"
    >
      <div
        class="w-full flex-col border-slate-200 max-md:hidden sm:hidden lg:flex border-b"
      >
        <div
          class="flex w-full items-center justify-between px-6 py-3"
        >
          <h2
            class="font-medium"
          >
            test-chat-session
          </h2>
          <button
            class="inline-flex items-center justify-center text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground rounded-full h-8 w-8"
            data-state="closed"
          >
            <svg
              class="lucide lucide-zap"
              fill="none"
              height="16"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="13 2 3 14 12 14 11 22 21 10 12 10 13 2"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="flex h-0 flex-1"
      >
        <div
          class="flex flex-1 flex-col"
        >
          <div
            class="flex flex-1 flex-col gap-12 overflow-auto pb-24 pl-6 pr-18 pt-6 scrollbar-none"
          >
            <div
              class="flex flex-col "
            >
              <div
                class="flex gap-2 "
              >
                <div
                  class="relative h-14 w-14"
                >
                  <span
                    class="flex shrink-0 overflow-hidden rounded-full relative h-12 w-12 bg-sky-600"
                  >
                    <span
                      class="flex h-full w-full items-center justify-center rounded-full bg-transparent text-white"
                    />
                  </span>
                </div>
                <div
                  class="flex flex-col"
                >
                  <div
                    class="mb-5 flex items-center gap-1"
                  >
                    <p
                      class="inline-block text-sm font-medium text-black"
                    >
                      Me
                    </p>
                    <p
                      class="inline-block text-xs font-medium leading-5 text-gray-500"
                    >
                      01/01/2023 12:00 AM
                    </p>
                  </div>
                  <div
                    class="flex items-end"
                  >
                    <div
                      class="relative max-w-[280px] rounded-lg p-4 sm:max-w-xs md:max-w-lg lg:max-w-3xl xl:max-w-3xl bg-primary"
                    >
                      <div
                        class="prose break-words dark:prose-invert prose-p:leading-relaxed prose-pre:p-0 text-white"
                      >
                        <p
                          class="prose-sm mb-2 font-normal last:mb-0"
                        >
                          hello1
                        </p>
                      </div>
                      <div
                        class="absolute bottom-full left-full flex -translate-x-14 translate-y-4"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-col "
            >
              <div
                class="flex gap-2 "
              >
                <div
                  class="relative h-14 w-14"
                >
                  <span
                    class="flex shrink-0 overflow-hidden rounded-full relative h-12 w-12 bg-sky-600"
                  >
                    <span
                      class="flex h-full w-full items-center justify-center rounded-full bg-transparent text-white"
                    >
                      C
                    </span>
                  </span>
                </div>
                <div
                  class="flex flex-col"
                >
                  <div
                    class="mb-5 flex items-center gap-1"
                  >
                    <p
                      class="inline-block text-sm font-medium text-black"
                    >
                      chat-app
                    </p>
                    <p
                      class="inline-block text-xs font-medium leading-5 text-gray-500"
                    >
                      01/01/2023 12:00 AM
                    </p>
                  </div>
                  <div
                    class="flex items-end"
                  >
                    <div
                      class="relative max-w-[280px] rounded-lg p-4 sm:max-w-xs md:max-w-lg lg:max-w-3xl xl:max-w-3xl bg-gray-100"
                    >
                      <div
                        class="prose break-words dark:prose-invert prose-p:leading-relaxed prose-pre:p-0 text-black"
                      >
                        <p
                          class="prose-sm mb-2 font-normal last:mb-0"
                        >
                          hello2
                        </p>
                      </div>
                      <div
                        class="absolute bottom-full left-full flex -translate-x-14 translate-y-4"
                      >
                        <div
                          class="flex w-[4.75rem] flex-row-reverse"
                        >
                          <div
                            class="mr-1 rounded-md border bg-white p-2 cursor-pointer hover:bg-slate-100 active:bg-slate-300"
                          >
                            <svg
                              class="stroke-slate-400"
                              fill="none"
                              height="16"
                              stroke="currentColor"
                              stroke-linecap="round"
                              stroke-linejoin="round"
                              stroke-width="3"
                              viewBox="0 0 24 24"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M7 10v12"
                              />
                              <path
                                d="M15 5.88 14 10h5.83a2 2 0 0 1 1.92 2.56l-2.33 8A2 2 0 0 1 17.5 22H4a2 2 0 0 1-2-2v-8a2 2 0 0 1 2-2h2.76a2 2 0 0 0 1.79-1.11L12 2h0a3.13 3.13 0 0 1 3 3.88Z"
                              />
                            </svg>
                          </div>
                          <div
                            class="mr-1 rounded-md border bg-white p-2 cursor-pointer hover:bg-slate-100 active:bg-slate-300"
                          >
                            <svg
                              class="stroke-slate-400"
                              fill="none"
                              height="16"
                              stroke="currentColor"
                              stroke-linecap="round"
                              stroke-linejoin="round"
                              stroke-width="3"
                              viewBox="0 0 24 24"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M17 14V2"
                              />
                              <path
                                d="M9 18.12 10 14H4.17a2 2 0 0 1-1.92-2.56l2.33-8A2 2 0 0 1 6.5 2H20a2 2 0 0 1 2 2v8a2 2 0 0 1-2 2h-2.76a2 2 0 0 0-1.79 1.11L12 22h0a3.13 3.13 0 0 1-3-3.88Z"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-col "
            >
              <div
                class="flex gap-2 "
              >
                <div
                  class="relative h-14 w-14"
                >
                  <span
                    class="flex shrink-0 overflow-hidden rounded-full relative h-12 w-12 bg-sky-600"
                  >
                    <span
                      class="flex h-full w-full items-center justify-center rounded-full bg-transparent text-white"
                    />
                  </span>
                </div>
                <div
                  class="flex flex-col"
                >
                  <div
                    class="mb-5 flex items-center gap-1"
                  >
                    <p
                      class="inline-block text-sm font-medium text-black"
                    >
                      Me
                    </p>
                    <p
                      class="inline-block text-xs font-medium leading-5 text-gray-500"
                    >
                      01/01/2023 12:00 AM
                    </p>
                  </div>
                  <div
                    class="flex items-end"
                  >
                    <div
                      class="relative max-w-[280px] rounded-lg p-4 sm:max-w-xs md:max-w-lg lg:max-w-3xl xl:max-w-3xl bg-primary"
                    >
                      <div
                        class="prose break-words dark:prose-invert prose-p:leading-relaxed prose-pre:p-0 text-white"
                      >
                        <p
                          class="prose-sm mb-2 font-normal last:mb-0"
                        >
                          hello3
                        </p>
                      </div>
                      <div
                        class="absolute bottom-full left-full flex -translate-x-14 translate-y-4"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-col "
            >
              <div
                class="flex gap-2 "
              >
                <div
                  class="relative h-14 w-14"
                >
                  <span
                    class="flex shrink-0 overflow-hidden rounded-full relative h-12 w-12 bg-sky-600"
                  >
                    <span
                      class="flex h-full w-full items-center justify-center rounded-full bg-transparent text-white"
                    >
                      C
                    </span>
                  </span>
                </div>
                <div
                  class="flex flex-col"
                >
                  <div
                    class="mb-5 flex items-center gap-1"
                  >
                    <p
                      class="inline-block text-sm font-medium text-black"
                    >
                      chat-app
                    </p>
                    <p
                      class="inline-block text-xs font-medium leading-5 text-gray-500"
                    >
                      01/01/2023 12:00 AM
                    </p>
                  </div>
                  <div
                    class="flex items-end"
                  >
                    <div
                      class="relative max-w-[280px] rounded-lg p-4 sm:max-w-xs md:max-w-lg lg:max-w-3xl xl:max-w-3xl bg-gray-100"
                    >
                      <div
                        class="prose break-words dark:prose-invert prose-p:leading-relaxed prose-pre:p-0 text-black"
                      >
                        <p
                          class="prose-sm mb-2 font-normal last:mb-0"
                        >
                          hello4
                        </p>
                      </div>
                      <div
                        class="absolute bottom-full left-full flex -translate-x-14 translate-y-4"
                      >
                        <div
                          class="flex w-[4.75rem] flex-row-reverse"
                        >
                          <div
                            class="mr-1 rounded-md border bg-white p-2 cursor-pointer hover:bg-slate-100 active:bg-slate-300"
                          >
                            <svg
                              class="stroke-slate-400"
                              fill="none"
                              height="16"
                              stroke="currentColor"
                              stroke-linecap="round"
                              stroke-linejoin="round"
                              stroke-width="3"
                              viewBox="0 0 24 24"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M7 10v12"
                              />
                              <path
                                d="M15 5.88 14 10h5.83a2 2 0 0 1 1.92 2.56l-2.33 8A2 2 0 0 1 17.5 22H4a2 2 0 0 1-2-2v-8a2 2 0 0 1 2-2h2.76a2 2 0 0 0 1.79-1.11L12 2h0a3.13 3.13 0 0 1 3 3.88Z"
                              />
                            </svg>
                          </div>
                          <div
                            class="mr-1 rounded-md border bg-white p-2 cursor-pointer hover:bg-slate-100 active:bg-slate-300"
                          >
                            <svg
                              class="stroke-slate-400"
                              fill="none"
                              height="16"
                              stroke="currentColor"
                              stroke-linecap="round"
                              stroke-linejoin="round"
                              stroke-width="3"
                              viewBox="0 0 24 24"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M17 14V2"
                              />
                              <path
                                d="M9 18.12 10 14H4.17a2 2 0 0 1-1.92-2.56l2.33-8A2 2 0 0 1 6.5 2H20a2 2 0 0 1 2 2v8a2 2 0 0 1-2 2h-2.76a2 2 0 0 0-1.79 1.11L12 22h0a3.13 3.13 0 0 1-3-3.88Z"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="relative flex w-full flex-col gap-4 px-6 pb-4"
          >
            <div
              class="absolute top-[-60px] flex w-full items-center justify-center"
            >
              <button
                class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 bg-white"
              >
                <svg
                  class="mr-2"
                  fill="none"
                  height="16"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
                  />
                  <path
                    d="M21 3v5h-5"
                  />
                  <path
                    d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
                  />
                  <path
                    d="M8 16H3v5"
                  />
                </svg>
                Regenerate response
              </button>
            </div>
            <form>
              <div
                class="flex justify-between space-x-2"
              >
                <textarea
                  class="flex w-full resize-none rounded-md border border-input bg-transparent px-3 py-2 text-sm ring-offset-background placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 min-h-[40px]"
                  data-testid="input"
                  placeholder="Type a message"
                />
                <button
                  class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
                  data-testid="send"
                  disabled=""
                  type="submit"
                >
                  Send
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
